programs/server/play.html

```diff
@@ -645,6 +648,32 @@
         }
     }
 
+    /// Pressing Tab in textarea will increase indentation.
+    /// But for accessibility reasons, we will fall back to tab navigation if the user already used Tab for that.
+
+    let user_prefers_tab_navigation = false;
+
+    [...document.querySelectorAll('input')].map(elem => {
+        elem.onkeydown = (e) => {
+            if (e.key == 'Tab') { user_prefers_tab_navigation = true; }
+        };
+    });
+
+    query_area.onkeydown = (e) => {
+        if (e.key == 'Tab' && !event.shiftKey && !user_prefers_tab_navigation) {
+            let elem = e.target;
+            let selection_start = elem.selectionStart;
+            let selection_end = elem.selectionEnd;
+
+            elem.value = elem.value.substring(0, elem.selectionStart) + '    ' + elem.value.substring(elem.selectionEnd);
+            elem.selectionStart = selection_start + 4;
+            elem.selectionEnd = selection_start + 4;
+
+            e.preventDefault();
+            return false;
+        }
```

Comment:
Maybe we can also remove indentation on Shift+Tab. But it becames complicated, we also may want to add indentation to selected lines (like sublime text or vscode), so, let's keep it simple. I added snippet just in case.
```suggestion
        }
        if (e.key == 'Tab' && event.shiftKey && !user_prefers_tab_navigation) {
            /// remove indentation at the beginning of the current line
            let elem = e.target;
            let selection_start = elem.selectionStart;
            let selection_end = elem.selectionEnd;

            let selected_lines_start = elem.value.lastIndexOf('\n', selection_start) + 1;
            if (elem.value.substring(selected_lines_start, selected_lines_start + 4) == '    ') {
                elem.value = elem.value.substring(0, selected_lines_start) + elem.value.substring(selected_lines_start + 4);
                elem.selectionStart = Math.max(selection_start - 4, selected_lines_start);
                elem.selectionEnd = Math.max(selection_end - 4, selected_lines_start);
            }
            e.preventDefault();
            return false;
        }
```

Reply:
I left Shift+Tab unchanged for accessibility concerns - if you cannot use the mouse at this moment of time, you still can escape from the textarea.